crates/storage/snapshot/tree.rs

```diff
@@ -0,0 +1,788 @@
+use std::{
+    collections::{HashMap, HashSet},
+    sync::{atomic::AtomicBool, Arc, RwLock},
+};
+
+use ethrex_common::{
+    types::{AccountState, BlockHash},
+    Address, Bloom, H256, U256,
+};
+use tracing::{debug, error, info};
+
+use crate::{api::StoreEngine, hash_address_fixed};
+
+use super::{difflayer::DiffLayer, disklayer::DiskLayer, error::SnapshotError};
+
+/// It consists of one persistent base
+/// layer backed by a key-value store, on top of which arbitrarily many in-memory
+/// diff layers are topped.
+///
+/// The memory diffs can form a tree with branching, but
+/// the disk layer is singleton and common to all.
+///
+/// The goal of a state snapshot is twofold: to allow direct access to account and
+/// storage data to avoid expensive multi-level trie lookups; and to allow sorted,
+/// cheap iteration of the account/storage tries for sync aid.
+#[derive(Clone, Debug)]
+pub struct SnapshotTree {
+    db: Arc<dyn StoreEngine>,
+    layers: Arc<RwLock<Layers>>,
+}
+
+#[derive(Debug, Clone)]
+pub enum Layer {
+    DiskLayer(Arc<DiskLayer>),
+    DiffLayer(Arc<RwLock<DiffLayer>>),
+}
+
+pub type Layers = HashMap<H256, Layer>;
+
+impl SnapshotTree {
+    pub fn new(db: Arc<dyn StoreEngine>) -> Self {
+        SnapshotTree {
+            db,
+            layers: Default::default(),
+        }
+    }
+
+    /// Rebuilds the tree, marking all current layers stale, creating a new base disk layer from the given root.
+    pub fn rebuild(&self, block_hash: BlockHash, state_root: H256) {
+        let mut layers = self.layers.write().unwrap();
+
+        for layer in layers.values() {
+            match layer {
+                Layer::DiskLayer(disk_layer) => disk_layer.mark_stale(),
+                Layer::DiffLayer(diff_layer) => diff_layer.write().unwrap().mark_stale(),
+            };
+        }
+
+        layers.clear();
```

Comment:
What is the purpose of marking all layers as stale if we are then clearing them?